File: python/sym_handler/handler.py
from pathlib import Path

from devtools import debug

from .models import SymLogEntry, SymLambdaResponse


def handle(event, context) -> dict:
    log = SymLogEntry.parse_obj(event)
    print("Got event:")
    debug(log)

    if log.event.type == "escalate":
        response = handle_escalate(log)
    elif log.event.type == "deescalate":
        response = handle_deescalate(log)
    else:
        response = SymLambdaResponse(body={"ok": False}, errors=["Unknown event type"])

    return response.dict()


def handle_escalate(log: SymLogEntry) -> SymLambdaResponse:
    if log.run.flow.slug != "my-lambda-flow":
        return SymLambdaResponse(body={"ok": False}, errors=["Unknown flow"])

    # Call your business logic, passing, for example, the username and the target
    # do_something(log.actor.username, log.fields.target.srn.slug)

    return SymLambdaResponse(body={"ok": True})


def handle_deescalate(log: SymLogEntry)-> SymLambdaResponse:
    # Call your business logic, passing, for example, the username and the target
    # do_something(log.actor.username, log.fields.target.srn.slug)

    ...

    return SymLambdaResponse(body={"ok": True})


if __name__ == "__main__":
    path = Path(__file__).parent.parent.parent / "payload.json"
    log = SymLogEntry.parse_file(path)
    handle(log, None)

File: typescript/sym_handler/handler.ts
import path from "path";
import fs from "fs";
import { Context } from "aws-lambda";
import { ParsedSRN } from "./models";
import type { SymLogEntry } from "./models";

export const handle = async (
  log: SymLogEntry,
  _ctx?: Context
): Promise<Object> => {
  console.log("Got Event: ", log);

  if (log.event.type == "escalate") {
    return handleEscalate(log);
  } else if (log.event.type == "deescalate") {
    return handleDeescalate(log);
  } else {
    return { body: { "ok": false}, errors: ["Unknown event type"] };
  }
};

const handleEscalate = async (log: SymLogEntry): Promise<Object> => {
  const flow = new ParsedSRN(log.run.flow);
  const target = new ParsedSRN(log.fields.target.srn);

  if (flow.slug != "my-lambda-flow") {
    return { body: { "ok": false}, errors: ["Unknown flow"] };
  }

  // Call your business logic, passing, for example, the username and the target
  // doSomething(log.actor.username, target.srn.slug);

  return { body: { "ok": true}, errors: [] };
};

const handleDeescalate = async (log: SymLogEntry): Promise<Object> => {
  const target = new ParsedSRN(log.fields.target.srn);

  // Call your business logic, passing, for example, the username and the target
  // doSomething(log.actor.username, target.srn.slug);

  return { body: { "ok": true}, errors: [] };
};

if (require.main === module) {
  const payload = path.join(__dirname, "..", "..", "payload.json");
  const data = fs.readFileSync(payload);
  const log: SymLogEntry = JSON.parse(data.toString());
  handle(log).then((res) => {
    console.log(res);
  });
}
